pxr/base/tf/fileUtils.cpp

```diff
@@ -94,6 +93,21 @@ Tf_HasAttribute(
         }
         return false;
     }
+
+    // Ignore reparse points on network volumes. They can't be resolved
+    // properly, so simply remove the reparse point attribute and treat
+    // if like a regular file/directory.
+    if ((attribs & FILE_ATTRIBUTE_REPARSE_POINT) != 0) {
```

Comment:
Hi @marktucker, thanks for creating this PR! 
I was testing this change with symlinks that point to existing and non-existent locations on another network and noticed some behavior that does not follow what is outlined in TfFileUtils.h. It seems that `PathIsNetworkPathW` does not register symlink paths as network paths and that setting `attribs &= ~FILE_ATTRIBUTE_REPARSE_POINT` can sometimes cause `(attribs & attributes) == expected` to return false such that `TfIsFile` no longer returns true when checking a reparse point on a network drive.

I'm proposing the change below to check if the resolved symlink path is a network path and explicitly returning a value based on what attribute is being checked against. Does this proposed change still capture your intention and solve the bug mentioned by the Houdini user?
```
    // Ignore reparse points on network volumes. They can't be resolved
    // properly, so simply remove the reparse point attribute and treat
    // if like a regular file/directory.
    if ((attribs & FILE_ATTRIBUTE_REPARSE_POINT) != 0) {
        // Calling PathIsNetworkPath sometimes sets the "last error" to an
        // error code indicating an incomplete overlapped I/O function. We
        // want to ignore this error.
        DWORD olderr = GetLastError();
        std::string resolvedPath = TfReadLink(path.c_str());
        if (PathIsNetworkPathW(ArchWindowsUtf8ToUtf16(resolvedPath).c_str())) {
            // Return false for checking if a path is a symlink but true for all
            // other cases
            return expected != FILE_ATTRIBUTE_REPARSE_POINT;
        }
        SetLastError(olderr);
    }
```

Reply:
Hi @Denise-Yang , thanks for looking into this. But I'm not sure I follow what you're saying... I'm having trouble conceiving of a concrete case where `TfIsFile` would return `false` with this PR for a path where this would have previously returned `true` (and where the correct answer is `true`). Could you provide a concrete example of how you set things up to return this incorrect result?

I can say that the impetus behind this PR is the fact that `TfReadLink`, when called on a reparse point on a network drive, returns a value that is completely meaningless on the local machine. So any use of the `resolvedPath` in your suggested code is not going to give a meaningful result. For example, suppose "S:" is a mounted network drive and "S:\path\to\some" is a reparse point to "D:\another\location". But "D" is drive "D" on the remote machine. Your code will then call `PathIsNetworkPathW("D:\another\location")`. `PathIsNetworkPathW` will look at drive "D" on the _local_ machine, which is _not_ the right drive "D" (the local machine may not have a drive D). That's why the value returned by `PathIsNetworkPathW` is meaningless, and so I don't think your change can remain as is, even if there is a bug in the original PR.

Reply:
Ah I see! I have realized I also misunderstood the intent of the PR thinking that we wanted to entirely ignore reparse points pointing _to_ network file systems (which in hindsight seems a little strange) and thus structured my test case accordingly.
I agree with your PR, and will work on testing it against a reparse point _on_ a network file system! Out of curiosity did you reproduce the bug using multiple Windows machines or were you able to find a workaround having multiple machines?

Reply:
Oh good, you had me worried there :)

I was able to reproduce the bug, but to do so I needed to have a separate "file server" machine and "test machine". If you use the same machine for the "server" and the "test machine", then the drive letters on both side of the network mount "match up". The reason this bug occurs is that the drive letters on the "server" differ from the drive letters on the "test machine".

Reply:
Hi Mark!
I was finally able to reproduce the bug :) (or at least what I think is a similar example), and this PR works as documented except in the case where the file/directory on the remote machine does not actually exist.

For my specific case, I had a symlink on a network share `T:\local\path` that points to `C:\remote\pathA\` that then points to `C:\remote\pathB\` where `C:\` was on a different machine. Evaluating `T:\local\path\directory\`, where _directory_ exists at`C:\remote\pathB\directory`, and `T:\local\path\` produce expected results until I deleted  `C:\remote\pathB\`, which then caused`GetFileAttributesW` to return the `error network path not found` when trying to evaluate `T:\local\path\directory\`.

Looking at the [docs for symlinks](https://learn.microsoft.com/en-us/windows/win32/fileio/symbolic-link-effects-on-file-systems-functions?redirectedfrom=MSDN), it seems that `GetFileAttributes` does not handle paths that contain symlinks. I'd like to confirm if this is a correct repro of the bug, and or if you've encountered a similar issue when making this PR (to my understanding, we should be handling this case)?

Reply:
Yep, that sounds like an accurate reproduction of the problematic situation.

Once you delete `C:\remote\pathB\`, I think it is correct for the `path not found` return for `T:\local\path\directory\`. Because indeed the requested path does not exist, right?

The known issue with this PR pointed out by @pmolodo was that for purely local symlinks, when the destination of the link gets deleted querying the source end of the symlink would return `path not found`. But with this PR, after deleting `C:\remote\pathB\`, querying `T:\local\path\` would return saying the path _does_ exist. Which is wrong according to the purely local equivalent case.

My argument for accepting this PR as-is in spite of this outstanding issue was:
1) Even if this one case is wrong, lots and lots of other cases that were wrong before will now work.
2) The harm of returning "this path exists" in this case is not relevant to the use of this function within USD. So in some other application using this method, we might care that this result is incorrect. But the functioning of the USD library will be unaffected by leaving this bug in place.
3) As far as we were able to determine, there is no sure fire way from the client end to detect if the destination of the symlink on the server exists. In other words Windows does not provide the capability to produce the truly "correct" answer in this scenario.

Reply:
Ah so just to clarify it's also expected for `T:\local\path\directory` is a file and is a directory queries to return that it is neither? 

I interpreted this discussion and the functions comments, "_even if resolveSymlinks is true we are unable to follow the symlink, and so will return true even if the destination of the symlink does not exist._", to mean that it should return that the path is both a file and directory if the destination of the symlink does not exist.

Reply:
The comment in the code refers only to the case where the specified path point _directly_ to the symlink. It does not apply to a query about any file or directory _under_ the symlink. Because the function being modified here doesn't pay any attention to symlinks part way down the specified path. When you call this function with the path `T:\local\path\directory`, the code here neither knows nor cares that `T:\local\path\` is a symlink.

I think @pmolodo was similarly misled (maybe by my wording in that comment), as referred to in the thread that ends with https://github.com/PixarAnimationStudios/OpenUSD/pull/2934#discussion_r1534663283.

Reply:
I see, thanks for all the clarifications Mark! Despite reading the thread, I still managed to misunderstand the comments to mean that `Tf_HasAttribute("//B/share/path/to/some/directory")` would now behave differently, but this makes a lot more sense!